# Report
## Project 2 - Continuous Control
### Udacity Deep Reinforcement Learning Nanodegree
by Kitson Swann

## Reacher Environment

In this environment, our goal is to use a Deep Reinforcement Learning Agent to train a double-jointed arm to 
move its hand to a target location.

For the purpose of this assignment, the task is considered solved when the agent achieves a score of +30 
over 100 consecutive episodes.

### States

The state space contains 33 dimensions corresponding to: 

- position rotation  
- velocity
- angular velocities of the arm 

### Actions

Each action is a vector with four numbers, corresponding to torque 
applicable to two joints. Every entry in the action vector should 
be a number between -1 and 1. 

### Rewards

According to the project instructions a reward of +0.1 is provided for each step that the agent's hand 
is in the goal location. However, in practice the rewards from the environment are much more varied ranging from 0 to 0.5. I am curious if the environment was updated between the time this assignment was authored and now. The different reward structure did not prevent me from succeeding at the task, but I wonder if it made it more difficult.  

## Agent Version 1

For this project I had planned to initially begin with the simple single agent environment and move onto the more complex multi-agent environment after solving the simple one, but achieving the minimum average score on the single agent version proved quite difficult, so I stopped at solving just the simple environment. Below you can see the agent initialized with the state and action spaces shown.

In [2]:
from unityagents import UnityEnvironment
import matplotlib.pyplot as plt
from optimize import find_optimal_hyperparameters
from train import train_ddpg

%matplotlib inline

env = UnityEnvironment(file_name='../env/v1/Reacher.app')

# get the default brain
brain_name = env.brain_names[0]
brain = env.brains[brain_name]

# reset the environment
env_info = env.reset(train_mode=True)[brain_name]

# number of agents
num_agents = len(env_info.agents)
print('Number of agents:', num_agents)

# size of each action
action_size = brain.vector_action_space_size
print('Size of each action:', action_size)

# examine the state space 
states = env_info.vector_observations
state_size = states.shape[1]
print('There are {} agents. Each observes a state with length: {}'.format(states.shape[0], state_size))
print('The state for the first agent looks like:', states[0])

INFO:unityagents:
'Academy' started successfully!
Unity Academy name: Academy
        Number of Brains: 1
        Number of External Brains : 1
        Lesson number : 0
        Reset Parameters :
		goal_size -> 5.0
		goal_speed -> 1.0
Unity brain name: ReacherBrain
        Number of Visual Observations (per agent): 0
        Vector Observation space type: continuous
        Vector Observation space size (per agent): 33
        Number of stacked Vector Observation: 1
        Vector Action space type: continuous
        Vector Action space size (per agent): 4
        Vector Action descriptions: , , , 


Number of agents: 1
Size of each action: 4
There are 1 agents. Each observes a state with length: 33
The state for the first agent looks like: [ 0.00000000e+00 -4.00000000e+00  0.00000000e+00  1.00000000e+00
 -0.00000000e+00 -0.00000000e+00 -4.37113883e-08  0.00000000e+00
  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
  0.00000000e+00  0.00000000e+00 -1.00000000e+01  0.00000000e+00
  1.00000000e+00 -0.00000000e+00 -0.00000000e+00 -4.37113883e-08
  0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
  0.00000000e+00  0.00000000e+00  5.75471878e+00 -1.00000000e+00
  5.55726671e+00  0.00000000e+00  1.00000000e+00  0.00000000e+00
 -1.68164849e-01]


## Solution

### Algorithm

As suggested in the project instructions I used the Deep Deterministic Policy Gradient agent from the Pendulum environment example [here](https://github.com/udacity/deep-reinforcement-learning/tree/master/ddpg-pendulum) as a starting point. I modified the training loop to work with the Unity Reacher environment, which required some minor changes from the original which was based on an OpenAI Gym environment.

### Initial Experimentation

After getting the training routine working, I initially had no success in getting the agent to train with the standard hyper-parameters. The score was oscillating around 0-0.1 most of the time and not improving even over 100 episodes.

Normally at this point I would have experimented with hyperparameter-optimization, but in this algorithm there are so many parameters that I felt the need to look for some good values. I consulted Udacity's Knowledge Base for some assistance, and saw a number of suggestions. I reviewed the suggestions and many repositories containing the solutions of other students for ideas. I made a number of changes as a result including:

- Changing the training loop to end when done which for the Reacher environment is 1000 timesteps.
- Tried different random seeds
- Modified the architecture of the actor and critic neural networks to try different hidden layer sizes.
- Modified the architecture of the actor and critic neural networks to add batch normalization.
- Modified the ddpg agent to only do a learning step every 5, 10, 20, 40 iterations.
- Reset the noise after every episode.
- Tried different suggested values for buffer size, batch size, gamma, tau, weight decay and the sigma variable in the Ornstein-Uhlenbeck process noise implementation.

Even with all of the above modifications I still wasn't seeing any decent results. The score rarely went above 0.5, and seemed to get worse over time.

### Structured Hyper-parameter Optimization

After seeing a number of people say the suggested changes should work, but having no success myself, I felt that I needed to take a more robust approach to hyper-parameter optimization. Similar to my implementation in Project 1, I used scikit-optimize's Gaussian Process Optimizer to search for a better set of hyper-parameters. Below is the search space that I initially defined.

I initially did batches of 10 episodes to keep the run time short and look for some better initial values. Because the process takes a long time I watched the values from each run, and visually tried to understand what values of each parameter were working well. As I noticed good values, I iteratively removed parameters from the search space, instead setting them at reasonable levels until I found something that worked fairly well which I've called Reasonable Parameter Set below.

I have included the call to the optimization routine below, but haven't run it as it takes up a lot of space, and I re-ran it iteratively so one single output was never the final one used.

In [ ]:
from skopt.space import Real, Integer, Categorical

space = [
    Real(0, 0.5, "uniform", name='eps_end'),
    Real(1e-5, 1e0, "log-uniform", name='eps_decay'),
    Categorical([1e5, 1e6], name="buffer_size"),
    Categorical([64, 128, 256, 512], name="batch_size"),
    Categorical([1, 5, 10, 20, 40], name="update_every"),
    Categorical([1, 5, 10, 20], name="update_times"),
    Categorical([128, 256, 512], name="actor_fc1_units"),
    Categorical([128, 256, 512], name="actor_fc2_units"),
    Categorical([128, 256, 512], name="critic_fc1_units"),
    Categorical([128, 256, 512], name="critic_fc2_units"),
    Real(1e-5, 1e-3, "log-uniform", name='actor_lr'),
    Real(1e-5, 1e-3, "log-uniform", name='critic_lr'),
    Real(0.9, 0.99, "uniform", name="gamma"),
    Real(1e-2, 1e-1, "log-uniform", name="tau"),
    Real(0, 1e-3, "uniform", name="weight_decay"),
    Real(0.01, 0.05, "log-uniform", name="noise_theta"),
    Real(0.01, 0.05, "log-uniform", name="noise_sigma"),
]

params, res_gp = find_optimal_hyperparameters(env=env, brain_name=brain_name, num_agents=num_agents, n_calls=50, episodes_per_batch=10, space=space)

#### Reasonable Parameter Set

This is the best resonable parameter set I found through a somewhat manual search for hyper-parameters. There are many parameters to tune here, and the search space grows so quickly that a full optimization run was going to take more time than I had available, so I viewed the results and manually tuned based on what I saw to develop the set of values below.

In [3]:
optimal_params = {
  "random_seed": 22,
  "noise_sigma": 0.02,
  "noise_theta": 0.015,
  "weight_decay": 0,
  "tau": 0.05,
  "gamma": 0.95,
  "critic_lr": 0.0001,
  "actor_lr": 0.001,
  "critic_fc2_units": 128,
  "critic_fc1_units": 128,
  "actor_fc2_units": 128,
  "actor_fc1_units": 128,
  "update_times": 10,
  "update_every": 20,
  "batch_size": 256,
  "buffer_size": 1000000.0,
  "eps_decay": 0.7,
  "eps_end": 0.3,
  "eps_start": 1.0,
  "action_size": 4,
  "state_size": 33,
  "n_episodes": 2000,
  "solved_threshold": 31.0,
  "break_early": True,
  "name": "_optimal",
  "num_agents": 1,
  "brain_name": "ReacherBrain"
}

scores = train_ddpg(env=env, **optimal_params)

{
  "random_seed": 22,
  "noise_sigma": 0.02,
  "noise_theta": 0.015,
  "weight_decay": 0,
  "tau": 0.05,
  "gamma": 0.95,
  "critic_lr": 0.0001,
  "actor_lr": 0.001,
  "critic_fc2_units": 128,
  "critic_fc1_units": 128,
  "actor_fc2_units": 128,
  "actor_fc1_units": 128,
  "update_times": 10,
  "update_every": 20,
  "batch_size": 256,
  "buffer_size": 1000000.0,
  "eps_decay": 0.7,
  "eps_end": 0.3,
  "eps_start": 1.0,
  "action_size": 4,
  "state_size": 33,
  "n_episodes": 2000,
  "solved_threshold": 31.0,
  "break_early": true,
  "name": "_optimal",
  "num_agents": 1,
  "brain_name": "ReacherBrain"
Replay Buffer Size: 1000000.0, Batch Size: 256


/Users/kitson_swann/Documents/udacity_drl/udacity_drl_p2/src/ddpg_agent.py:140: UserWarning: torch.nn.utils.clip_grad_norm is now deprecated in favor of torch.nn.utils.clip_grad_norm_.
  torch.nn.utils.clip_grad_norm(self.critic_local.parameters(), 1)


Episode 100	Average Score: 1.69
Episode 200	Average Score: 7.48
Episode 300	Average Score: 13.16
Episode 400	Average Score: 16.45
Episode 500	Average Score: 23.23
Episode 600	Average Score: 29.82
Episode 639	Average Score: 31.07
Environment solved in 639 episodes!	Average Score: 31.07


In [6]:
import json

with open("scores.json", "w") as f:
    json.dump({"scores": scores}, f, indent=2)

### Success !!!

As you can see above the environment was solved in 639 episodes. In reality, it was a bit shorter than that, because I set the threshold to 31 just to go a bit beyond the minimim value.

Below you can see the plots of raw scores and average scores that I took from Tensorboard.

![tensorboard_raw_score.png](tensorboard_raw_score.png)
![tensorboard_avg_score.png](tensorboard_avg_score.png)

## Watch The Trained Agent

In [ ]:
from unityagents import UnityEnvironment
env = UnityEnvironment(file_name='../env/v1/Reacher.app')

# get the default brain
brain_name = env.brain_names[0]
brain = env.brains[brain_name]
from ddpg_agent import Agent
import torch
import numpy as np

params = {
  "random_seed": 22,
  "noise_sigma": 0.02,
  "noise_theta": 0.015,
  "weight_decay": 0,
  "tau": 0.05,
  "gamma": 0.95,
  "critic_lr": 0.0001,
  "actor_lr": 0.001,
  "critic_fc2_units": 128,
  "critic_fc1_units": 128,
  "actor_fc2_units": 128,
  "actor_fc1_units": 128,
  "update_times": 10,
  "update_every": 20,
  "batch_size": 256,
  "buffer_size": 1000000.0,
  "eps_decay": 0.7,
  "eps_end": 0.3,
  "eps_start": 1.0,
  "action_size": 4,
  "state_size": 33
}

agent = Agent(**params)

agent.actor_local.load_state_dict(torch.load('checkpoint_actor_optimal.pth'))
agent.critic_local.load_state_dict(torch.load('checkpoint_critic_optimal.pth'))
        
env_info = env.reset(train_mode=False)[brain_name]
agent.reset()
states = env_info.vector_observations
agent_scores = np.zeros(1)
while True:
    actions = agent.act(state=states)
    env_info = env.step(actions)[brain_name]
    next_states = env_info.vector_observations
    rewards = env_info.rewards
    dones = env_info.local_done
    if np.any(dones):
        break

env.close()